[DR-3148] Support slashes in DRS Object Ids

nmalfroy opened the pull request:
This PR allows users to resolve drs object ids with slashes in them (this is a requirement of working with Azure drs ids).  E.g. this means that a user could call:
```
POST /api/v4/drs/resolve
{
  "url": "drs://jade.datarepo-dev.broadinstitute.org/foo/bar"
}
```
or
```
POST /api/v4/drs/resolve
{
  "url": "drs://jade.datarepo-dev.broadinstitute.org/foo%2Fbar"
}
```

and that will turn into DrsHub calling:
```
GET https://jade.datarepo-dev.broadinstitute.org/ga4gh/drs/v1/objects/foo%2Fbar
```

It's accomplished by updating how the client (only for ga4gh) is generated so that it eagerly encodes path/query parameters (even though the later isn't really needed, but forward proofing) and then doesn't re-encodes when generating the final URL to use to resolve DrsObjects.  It also eagerly.

It also attempts to decode object ids before sending them to be resolved.  I agree this is a strange thing but it lines up with how Martha works.

The testing is done via a modification to one of the existing integration tests and unit testing for the objectId decoding.

This is in conjunction with https://github.com/DataBiosphere/jade-data-repo/pull/1485 and brings this behavior to parity with Martha

Review note: commit-by-commit will make this much easier to review (e.g. I isolated the base template introduction commit and the changes).

nmalfroy committed:
Add base codegen template

nmalfroy committed:
Update mustache templates to explicitly encode path and query parameters when building up requests

...and don't encode when full encoding the URL at the end.  This allows us to isolate the encoding

nmalfroy committed:
Make sure that the update codegen only gets used for the ga4gh client

nmalfroy committed:
Add an integration test to make sure aliases and slashes work

nmalfroy committed:
Decode objectIds before attempting to fetch the object

tlangs commented on integration/src/main/java/scripts/testscripts/ResolveTdrDrs.java:
```diff
@@ -6,38 +6,48 @@
+      List.of(
+          "drs://jade.datarepo-dev.broadinstitute.org/v1_93dc1e76-8f1c-4949-8f9b-07a087f3ce7b_8b07563a-542f-4b5c-9e00-e8fe6b1861de",
+          "drs://jade.datarepo-dev.broadinstitute.org/drshubfoo",
+          "drs://jade.datarepo-dev.broadinstitute.org/drshubfoo/bar");
```
Did we want to add a url containing `%2F` here as a test? Also, we should talk about Pact testing between DRSHub and TDR!

nmalfroy replied:
Ah good catch.  I'll add that

tlangs reviewed (changes requested):
Could we unit-test the generated code? The generated `ApiClient` should have an `ApiClient(RestTemplate restTemplate)` constructor, and we could pass a mock `RestTemplate` into that and make sure that the URL Encoding is happening as expected. Would also let us test against new versions of the DRS API spec and swagger codegen upgrades.

nmalfroy commented:
I was trying to find a way to test it given the existing framework and it didn't seem like a thing that was practical (e.g. we need a handle on the low level HTTP client and I don't think the client is factored in a way that lets us get that).  That's why I went the integration test route to test that it does the right thing but it's an interesting idea to try with the resttemplate mocking. I'll see what I can do

nmalfroy committed:
Add encoded Drs URI in the integration test

nmalfroy committed:
Make sure that objectIds are properly decode/re-encoded in  all cases

nmalfroy committed:
Mock out the lower-level RestTemplate for unit testing of path encoding

nmalfroy commented:
Ok @tlangs I updated the unit tests I added to use a mocked out rest template (I left the existing unit tests alone since mocking that out would get real unwieldy)
